dev_docs/developer_guidelines.md

```diff
@@ -0,0 +1,85 @@
+# Developer guidelines
+
+### Requirements
+- npm
+- Python 3.6+
+- Chrome
+
+[See dev section of README](../README.md)
+
+## Server Dev
+### Install
+* Build the client and put static files in place: `make build-for-server-dev`
+* Install from local files: `make install-dev`
+
+### Launch
+* `cellxgene launch [options] <datafile>`
+
+### Reloading
+If you install cellxgene using `make install-dev` the server will be restarted every time you make changes on the server code. If changes affects the client, the browser must be reloaded.
+
+### Linter
+We use `flake8` to lint code. Travis CI runs `flake8 server`.
+
+### Test
+1. Install development requirements `pip install -r server/requirements-dev.txt`
+2. Run tests `pytest server/test`
+
+### Tips
+* Install in a virtualenv
+* May need to rebuild/reinstall when you make client changes
+
+## Client Dev
+### Install
+1. Install prereqs for client: `npm install --prefix client/ client`
+2. Install cellxgene server: `pip install -e .` Caveat: this will install necessary files for launching the cellxgene REST API but, you will need to build the client if you want to be able to view cellxgene in the browser when launched from the CLI.
+
+### Launch
+To launch with hot reloading you need to launch the server and the client separately. Node's hot reloading starts the client on its own node server and auto-refreshes when changes are made.
+1. Launch server (the client relies on the REST API being available): `cellxgene launch [options] <datafile>`
+2. Launch client: `npm run start`
+3. Client will be served on localhost:3000
+
+### Build
+To build only the client: `make build-client`
+
+### Linter
+We use `prettier` to lint the code.
+
+### Test
+Run `npm run unit-test`
+
+### Tips
+* You can also install/launch the server side code from npm scrips (requires python3.6 with virtualenv) `npm run backend-dev`
+
+## Running tests
+Client and server tests run on Travis CI for every push, PR, and commit to master on github. Smoke tests run nightly on master only. 
+
+### Server
+Install development requirements `pip install -r server/requirements-dev.txt`
+Run tests `pytest server/test`
```

Comment:
btw, if you're not using the parallel version of pytest, you may want to consider it. :)

Reply:
@csweaver - this seems like an excellent idea give how long it takes to run our tests....

Reply:
pytest-xdist is your friend.

Reply:
So it's would be a little complicated considering some of our tests launch the server there might be port collisions. (Not extremely complicated, just not trivial).

As for the speed of the tests, they actually only take ~13 seconds for all of them. The reason the travis build takes so long is building the client. So for true speed improvements we should look to webpack 😁

Reply:
file the webpack issue and assign to me - happy to work on it!